File: src/main/GO/leetcode/lists/hot100/LeetCode297.go
package hot100

import (
	"strconv"
	"strings"
)

//https://leetcode.com/problems/serialize-and-deserialize-binary-tree/solutions/552538/golang-with-queue/

type Codec struct {
}

func ConstructorCodec() Codec {
	return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	if root == nil {
		return "X,"
	}
	left := this.serialize(root.Left)
	right := this.serialize(root.Right)

	return strconv.Itoa(root.Val) + "," + left + right
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	nodes := strings.Split(data, ",")
	return this.deserializeHelper(&nodes)
}

func (this *Codec) deserializeHelper(nodes *[]string) *TreeNode {
	nodeVal := "X"
	if len(*nodes) > 0 {
		nodeVal = (*nodes)[0]
		*nodes = (*nodes)[1:]
	}
	if nodeVal == "X" {
		return nil
	}
	val, _ := strconv.Atoi(nodeVal)
	root := TreeNode{Val: val}
	root.Left = this.deserializeHelper(nodes)
	root.Right = this.deserializeHelper(nodes)

	return &root

}

File: src/main/java/leetcode/lists/hot100/LeetCode297.java
package leetcode.lists.hot100;

import common.TreeNode;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/7 23:40
 */
public class LeetCode297 {

    static class Codec {

        public String serialize(TreeNode root) {
            return rserialize(root, "");
        }

        public TreeNode deserialize(String data) {
            String[] dataArray = data.split(",");
            List<String> dataList = new LinkedList<>(Arrays.asList(dataArray));
            return rdeserialize(dataList);
        }

        public String rserialize(TreeNode root, String str) {
            if (root == null) {
                str += "None,";
            } else {
                str += root.val + ",";
                str = rserialize(root.left, str);
                str = rserialize(root.right, str);
            }
            return str;
        }

        public TreeNode rdeserialize(List<String> dataList) {
            if (dataList.get(0).equals("None")) {
                dataList.remove(0);
                return null;
            }
            TreeNode root = new TreeNode(Integer.parseInt(dataList.get(0)));
            dataList.remove(0);
            root.left = rdeserialize(dataList);
            root.right = rdeserialize(dataList);
            return root;
        }
    }
}
